### PR TITLE
Add QAT Walkthrough Notebook example

### PR DESCRIPTION
**Pull request by farshadghodsian**
## What does this PR do?

**Type of change:** New example 

**Overview:**  
Adding a QAT Jupyter Notebook example that walks user through how to apply Quantization Aware Training (QAT) to an LLM, Meta's Llama-3.1-8b, and serve it via TensorRT-LLM Docker container.

## Usage
See new QAT Walkthrough Notebook in `./examples/llm_qat/notebooks` for usage instructions

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes 
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: No 

## Summary by CodeRabbit

* **New Features**
  * End-to-end Quantization Aware Training walkthrough for large models: NVFP4 calibration, quantization, QAT training, checkpointing, export, and TensorRT-LLM deployment with example inference.

* **Documentation**
  * Step-by-step notebook covering prerequisites, model/dataset setup, training/calibration/quantization workflow, sample outputs, config notes, and Docker-based deployment/serving with an example request.

* **Chores**
  * Added notebook dependencies: ipywidgets, nvidia-modelopt[all], and trl.

### CODE REVIEW

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+   "source": [
+    "trainer.train()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f9741ead-937c-42a8-9c06-0cb547605ebf",
+   "metadata": {},
+   "source": [
+    "# Deploying the QAT Model with TensorRT-LLM"
```
Comment (realAsma) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: The deployment path in this notebook is not correct. The model after QAT is fake quantized and quantization happens in Pytorch. This needs to be exported to real quantized model for deployment with TRTLLM/vLLM/SGLang as explained in https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/cb98b221e1b1730226257e20b4c81ebb259fc2d6/examples/llm_qat#deployment

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+   "source": [
+    "# NVIDIA ModelOpt Quantization Aware Training (QAT) Walkthrough"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a695be45-1472-42bc-824e-5c992a487fa7",
+   "metadata": {},
+   "source": [
+    "**Quantization Aware Training (QAT)** is a method that simulates the effects of quantization during neural network post-training to preserve accuracy when deploying models in very-low-precision formats. Unlike post-training quantization, QAT inserts \"fake quantization\" nodes into the computational graph, mimicking the rounding and clamping operations that occur during actual quantization. This allows the model to adapt its weights and activations to mitigate accuracy loss.\n",
```
Comment (RalphMao) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]:
```suggestion
    "**Quantization Aware Training (QAT)** is a method that learns the effects of quantization during neural network training to preserve accuracy when deploying models in very-low-precision formats. QAT inserts quantizer nodes into the computational graph, mimicking the rounding and clamping operations that occur during actual quantization. This allows the model to adapt its weights and activations to mitigate accuracy loss.\n",
```

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+    "model_kwargs = {\n",
+    "    \"revision\": model_args.model_revision,\n",
+    "    \"trust_remote_code\": model_args.trust_remote_code,\n",
+    "    \"attn_implementation\": model_args.attn_implementation,\n",
+    "    \"torch_dtype\": model_args.torch_dtype,\n",
+    "    \"use_cache\": False,\n",
+    "    \"device_map\": \"auto\",\n",
+    "}\n",
+    "\n",
+    "# Dequantize if the model is in MXFP4 format (for gpt-oss family of models)\n",
```
Comment (RalphMao) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]:
```suggestion
```

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 1 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: Updated to add HF Checkpoint conversion steps.

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 2 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: This has been fixed. Removed reference to fake quantization. Thanks for the suggestion.

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 3 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: Removed this section and any reference to GPT-OSS since that model has it's own example notebook. Also the specific MXFP4 config is only available in the nightly build  of Model Opt currently

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "0f8af903-cde2-4fcd-b725-044876b73e8a",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
```
Comment (realAsma) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: The model does not seem to be trained. The starting loss is similar to the final loss. This might confuse users.

Probably we need to train longer. Could you train longer? Looks like 0.1 epochs took 2 minutes. Can we train for the entire 1 epoch?

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a695be45-1472-42bc-824e-5c992a487fa7",
+   "metadata": {},
+   "source": [
+    "**Quantization Aware Training (QAT)** is a method that learn the effects of quantization during neural network post-training to preserve accuracy when deploying models in very-low-precision formats. QAT inserts quantizer nodes into the computational graph, mimicking the rounding and clamping operations that occur during actual quantization. This allows the model to adapt its weights and activations to mitigate accuracy loss.\n",
+    "\n",
+    "This notebook demonstrates how to apply Quantization Aware Training (QAT) to an LLM, Meta's Llama-3.1-8b in this case with NVIDIA's TensorRT Model Optimizer (ModelOpt) QAT toolkit. We walk through downloading and loading the model, calibrating it using an example dataset, specifically CNN/DailyMail dataset sample, applying NVFP4 quantization, generating outputs, and exporting the quantized model."
```
Comment (realAsma) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: nit:
```suggestion
    "This notebook demonstrates how to apply Quantization Aware Training (QAT) to an LLM, Meta's Llama-3.1-8b in this case with NVIDIA's TensorRT Model Optimizer (ModelOpt) QAT toolkit. We walk through downloading and loading the model, calibrating it using an example dataset, specifically CNN/DailyMail dataset sample, applying NVFP4 quantization, generating outputs, and finally deploying the quantized model to TensorRT-LLM."
```

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+   "execution_count": 4,
+   "id": "b6af94af-1de6-4cb1-959b-98fb3f4e1932",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from transformers import AutoConfig\n",
+    "from trl import ModelConfig\n",
+    "\n",
+    "model_args = ModelConfig(\n",
+    "    model_name_or_path=\"meta-llama/Llama-3.1-8B-Instruct\",\n",
```
Comment (realAsma) [inline examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: nit:
Llama3 8B is an older model. Can we use `Qwen/Qwen3-8B`?
```suggestion
    "    model_name_or_path=\"Qwen/Qwen3-8B\",\n",
```

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 4 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: Ran a full epoch of training with Qwen3-8b which led to much better final loss numbers.

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 5 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: Updated the above as per your suggestion.

File: examples/llm_qat/notebooks/QAT_Walkthrough.ipynb (hunk 6 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_Walkthrough.ipynb]: Updated the Notebook to use Qwen3-8b model as per your suggestion.

**Comment by kevalmorabia97** [comment]
@farshadghodsian seems like your commits are not signed with an SSH key. Please take a look at CONTRIBUTING.md for steps to set it up and amend your commits to be signed before merging

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 7]
```diff
@@ -0,0 +1,844 @@
...
+   "source": [
+    "If you haven't already, install the required dependencies for this notebook. Key dependancies include:\n",
+    "\n",
+    "- nvidia-modelopt\n",
+    "- torch\n",
+    "- transformers\n",
+    "- jupyterlab\n",
+    "\n",
+    "This repo contains a `examples/llm_qat/notebooks/requirements.txt` file that can be used to install all required dependancies."
+   ]
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb:42]: Ignoring this recommendation.

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 8]
```diff
@@ -0,0 +1,844 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "eb2071be-df85-4961-92b0-567830a37d71",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "%env HF_TOKEN=<YOUR_HUGGINGFACE_TOKEN>"
+   ]
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb]: Thank you, but ignoring this suggestion.

File: examples/llm_qat/notebooks/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+ipywidgets
+nvidia-modelopt[all]
+torch
+torchvision
+transformers
```
Comment (kevalmorabia97) [inline examples/llm_qat/notebooks/requirements.txt]:
```suggestion
ipywidgets
nvidia-modelopt[hf]
```

File: examples/llm_qat/notebooks/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+trl
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/requirements.txt]: Ignoring this suggestion and going to switch to a shorter requirements.txt list as suggested by @kevalmorabia97

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 11]
```diff
@@ -0,0 +1,844 @@
...
+    "from trl import SFTTrainer\n",
+    "\n",
+    "trainer = SFTTrainer(\n",
+    "    model=model,\n",
+    "    args=training_args,\n",
+    "    train_dataset=dataset[script_args.dataset_train_split],\n",
+    "    eval_dataset=dataset[script_args.dataset_test_split],\n",
+    "    processing_class=tokenizer,\n",
+    ")"
+   ]
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb:354]: Code works without specifying a tokenizer if the tokenizer is already present in the model directory or in the config in the model directory so this is not needed.

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 12]
```diff
@@ -0,0 +1,844 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "6d25c2b1-a68b-4748-ac29-e8a893ce1762",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_name = \"Qwen/Qwen3-8B\""
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb]: Fixed this to remove mention of LLama mode. Thanks

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 13]
```diff
@@ -0,0 +1,844 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "a695be45-1472-42bc-824e-5c992a487fa7",
+   "metadata": {},
+   "source": [
+    "**Quantization Aware Training (QAT)** is a method that learn the effects of quantization during neural network post-training to preserve accuracy when deploying models in very-low-precision formats. QAT inserts quantizer nodes into the computational graph, mimicking the rounding and clamping operations that occur during actual quantization. This allows the model to adapt its weights and activations to mitigate accuracy loss.\n",
+    "\n",
+    "This notebook demonstrates how to apply Quantization Aware Training (QAT) to an LLM, Qwen3-8b in this case with NVIDIA's TensorRT Model Optimizer (ModelOpt) QAT toolkit. We walk through downloading and loading the model, calibrating it using an example dataset, specifically CNN/DailyMail dataset sample, applying NVFP4 quantization, generating outputs, and finally deploying the quantized model to TensorRT-LLM."
+   ]
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb]: Tweaked the description of the notebook based on suggestion.

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 14]
```diff
@@ -0,0 +1,844 @@
...
+   "source": [
+    "## Stop the TensorRT-LLM Docker contrainer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "acaeaaf3-fe9d-4592-9a5d-fee9721e507d",
+   "metadata": {},
+   "source": [
+    "Finally clean up the TensorRT-LLM server by stopping and exiting out of the containerthe Docker image. Alternatively you can run the below cell to stop the running container. The container should automatically delete itself once stopped as it was started with the `-rm` flag."
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb]: Good catch, fixed this.

File: examples/llm_qat/notebooks/requirements.txt (hunk 9 above)
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/requirements.txt]: Shortened the requirements.txt file as per your suggestion. I assume `nvidia-modelopt[hf]` package will pull in transformers and torch pre-reqs without having to specify them.

File: examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb [hunk 15]
```diff
@@ -0,0 +1,844 @@
...
+    "dataset = torch.utils.data.Subset(\n",
+    "    trainer.eval_dataset, list(range(min(len(trainer.eval_dataset), calib_size)))\n",
+    ")\n",
+    "data_loader = trainer.get_eval_dataloader(dataset)\n",
+    "\n",
+    "\n",
+    "def forward_loop(model):\n",
+    "    for data in data_loader:\n",
+    "        model(**data)"
+   ]
```
[the remark this answers is not in the document]
Comment (farshadghodsian) [reply examples/llm_qat/notebooks/QAT_QAD_Walkthrough.ipynb:390]: Unable to test this so ignoring for now.

**Comment by kevalmorabia97** [comment]
@farshadghodsian your commits are still not verified with an ssh key. Please refer to the steps here: https://github.com/NVIDIA/TensorRT-Model-Optimizer?tab=contributing-ov-file#%EF%B8%8F-signing-your-work

**Comment by farshadghodsian** [comment]
> @farshadghodsian your commits are still not verified with an ssh key. Please refer to the steps here: https://github.com/NVIDIA/TensorRT-Model-Optimizer?tab=contributing-ov-file#%EF%B8%8F-signing-your-work

I see my issue now. I forgot to add my signing key to my Github account. My commits should now be verified. ✅

**Review by kevalmorabia97 — APPROVED** [review]

**Comment by kevalmorabia97** [comment]
Code quality checks are failing:
https://github.com/NVIDIA/TensorRT-Model-Optimizer/actions/runs/17557846268/job/49867165107?pr=278
You can check CONTRIBUTING.md for steps to fix this

**Review by RalphMao — APPROVED** [review]
